### PR TITLE
Python: full tiktoken and HF tokenizers compat for release

### LINKED ISSUE #369: Python bindings: nice-to-have features for full tiktoken/tokenizers compat
## Summary

Tracking remaining gaps between wordchipper's Python compat layers and the upstream libraries. These are not blocking for a preview release but would improve drop-in compatibility.

## Tiktoken compat (`wordchipper.compat.tiktoken`)

- [ ] `encode_single_token(token: str) -> int`
- [ ] `decode_single_token_bytes(token_id: int) -> bytes`
- [ ] `encode_ordinary` should ignore special tokens (currently recognizes them as special token IDs; tiktoken encodes them as individual characters). See `test_encode_ordinary_special_diverges`.
- [ ] Default `disallowed_special="all"` should raise ValueError when special token text appears in input without `allowed_special`. Currently wordchipper's default encode doesn't recognize special tokens so the check never triggers. See `test_disallowed_special_diverges`.
- [ ] numpy array support for encode/decode return types

## HuggingFace tokenizers compat (`wordchipper.compat.tokenizers`)

- [ ] `pair` parameter in `encode()`/`encode_batch()` (sentence pair encoding)
- [ ] `is_pretokenized` parameter
- [ ] `add_special_tokens=False` support
- [ ] `skip_special_tokens=False` support in `decode()`/`decode_batch()`
- [ ] `with_added_tokens=False` in `get_vocab_size()`
- [ ] Extra kwargs in `from_pretrained()` (e.g., `revision`)

## Notes

All unimplemented features currently raise `NotImplementedError` with clear messages, which is the right UX for a preview. The encode/decode core paths match both tiktoken and HF tokenizers across all tested encodings and diverse input text (verified by side-by-side comparison tests).

The `encode_ordinary` and `disallowed_special` divergences are the highest priority items since they affect common tiktoken usage patterns.

### PR DESCRIPTION
**Pull request by antimora**
## Summary

Closes #369

Brings both compat layers to release-ready state. Side-by-side comparison tests verify parity with tiktoken and HF tokenizers across all encodings and diverse text inputs.

### Tiktoken compat

**New methods:**
- `encode_single_token(str|bytes)` / `decode_single_token_bytes(int)`
- `decode_bytes(tokens)` / `decode_bytes_batch(batch)` / `decode_tokens_bytes(tokens)`
- `encode_to_numpy()` (optional numpy dep)
- `is_special_token(token)`
- `token_byte_values()`

**Behavioral fixes:**
- `encode_ordinary`: uses `SpecialFilter.include_none()` so special token text is BPE-encoded as regular subwords
- `disallowed_special`: scans input text for special token strings before encoding (the old ID-based post-check never triggered because the default filter excludes specials)
- `decode()` / `decode_batch()`: `errors` param for UTF-8 handling (strict/ignore/replace)

### HF tokenizers compat

**Encoding fields:** `attention_mask`, `type_ids`, `special_tokens_mask`, `offsets`, `__len__()`, `__repr__()`

**New methods:**
- `get_vocab(with_added_tokens)`, `get_vocab_size(with_added_tokens=False)`
- `enable_padding()` / `no_padding()`, `enable_truncation()` / `no_truncation()`
- `padding` / `truncation` properties
- `num_special_tokens_to_add()`

**Behavioral fixes:**
- `pair` parameter in `encode()`/`encode_batch()` with `type_ids=1` for second sequence
- `add_special_tokens=False` accepted (no-op for GPT BPE)
- `skip_special_tokens` in `decode()`/`decode_batch()` filters special token IDs
- Extra kwargs in `from_pretrained()` accepted and ignored

### Not implemented (filed separately)
- `is_pretokenized` (#374), numpy array return types (#375)
- tiktoken: `encode_with_unstable`, `decode_with_offsets`, constructor (#376)
- HF: `from_file`, `save`, `add_tokens`, training (#377)

### Rust changes
- `_Tokenizer.decode_bytes()` / `decode_bytes_batch()` via existing `try_decode_to_bytes`
- `_Vocab.id_to_token_bytes()` returning `PyBytes`

## Test plan
- [x] 507 tests pass (up from 462)
- [x] Side-by-side tiktoken comparison across 4 encodings: encode_ordinary, decode_bytes, decode_tokens_bytes, decode_bytes_batch
- [x] Side-by-side HF comparison across 2 models: encode IDs, pair encoding, add_special_tokens=False
- [x] Encoding fields populated correctly (attention_mask, type_ids, special_tokens_mask)
- [x] Padding/truncation: fixed-length, left/right, pad_to_multiple_of
- [x] decode errors param, is_special_token, token_byte_values, get_vocab

### CODE REVIEW

File: bindings/python/py_src/wordchipper/compat/tiktoken.py [hunk 1]
```diff
@@ -257,25 +269,81 @@ def encode_batch(
...
+    def encode_single_token(self, text_or_bytes: str | bytes) -> int:
+        if isinstance(text_or_bytes, bytes):
+            try:
+                text_or_bytes = text_or_bytes.decode("utf-8")
+            except UnicodeDecodeError:
+                raise KeyError(text_or_bytes)
+        token_id = self._tok.vocab.token_to_id(text_or_bytes)
+        if token_id is None:
+            raise KeyError(text_or_bytes)
+        return token_id
```
[the remark this answers is not in the document]
Comment (antimora) [reply bindings/python/py_src/wordchipper/compat/tiktoken.py:301]: Valid point. The Rust `token_to_id` takes `&str` so we can't do a raw byte lookup today. We already catch `UnicodeDecodeError` and re-raise as `KeyError` (added in the prior review round) so the API contract matches tiktoken (callers get `KeyError` for any unrecognized input). A proper `token_bytes_to_id(bytes)` on the Rust side would fix the remaining gap for non-UTF-8 byte tokens.

File: bindings/python/py_src/wordchipper/compat/tokenizers.py [hunk 2]
```diff
@@ -67,36 +187,42 @@ def encode(
     ) -> Encoding:
         """Encode a string, returning an :class:`Encoding` with ids and tokens.
 
-        ``pair``, ``is_pretokenized``, and ``add_special_tokens`` are accepted
-        for API compatibility but raise :class:`NotImplementedError` if set to
-        non-default values.
+        ``is_pretokenized`` is accepted for API compatibility but raises
+        :class:`NotImplementedError` when set to ``True``.
         """
```
[the remark this answers is not in the document]
Comment (antimora) [reply bindings/python/py_src/wordchipper/compat/tokenizers.py:197]: Fixed. Updated the docstring to document `pair` (sets `type_ids=1` for second sequence) and `add_special_tokens` (accepted but no-op for GPT-BPE).

File: bindings/python/py_src/wordchipper/compat/tokenizers.py [hunk 3]
```diff
@@ -106,31 +232,99 @@ def decode(
...
+        self._truncation = None
+
+    @property
+    def truncation(self) -> dict | None:
+        return self._truncation
+
     # -- vocab inspection ----------------------------------------------------
 
+    def get_vocab(self, with_added_tokens: bool = True) -> dict[str, int]:
+        return self._tok.vocab.to_dict()
```
[the remark this answers is not in the document]
Comment (antimora) [reply bindings/python/py_src/wordchipper/compat/tokenizers.py]: Good catch. Fixed: `get_vocab(with_added_tokens=False)` now filters to tokens with IDs below `vocab_size`, consistent with how `get_vocab_size` handles the flag.

File: bindings/python/py_src/wordchipper/compat/tiktoken.py [hunk 4]
```diff
@@ -224,18 +220,34 @@ def encode(
...
+        return self._tok.encode(text, special_filter=allowed_filter)
+
+    def encode_to_numpy(
+        self,
+        text: str,
+        *,
+        allowed_special: Literal["all"] | AbstractSet[str] = frozenset(),
+        disallowed_special: Literal["all"] | Collection[str] = "all",
+    ):
+        import numpy as np
```
[the remark this answers is not in the document]
Comment (antimora) [reply bindings/python/py_src/wordchipper/compat/tiktoken.py]: Fixed. Now catches `ImportError` and raises with a clear message pointing to `pip install numpy`.

File: bindings/python/tests/compat/test_tokenizers.py [hunk 5]
```diff
@@ -192,52 +224,192 @@ def test_from_pretrained_bare_name(self):
...
         tok = self.get_tok()
-        with pytest.raises(NotImplementedError, match="skip_special_tokens"):
-            tok.decode_batch([[1, 2]], skip_special_tokens=False)
+        hello_ids = tok.encode("hello").ids
+        special_ids = list(tok._tok.specials.values())
+        mixed = hello_ids + special_ids
+        assert tok.decode(mixed, skip_special_tokens=True) == "hello"
+        # False includes the special token text
+        decoded_with = tok.decode(mixed, skip_special_tokens=False)
+        assert len(decoded_with) > len("hello")
```
[the remark this answers is not in the document]
Comment (antimora) [reply bindings/python/tests/compat/test_tokenizers.py]: Fixed. Simplified to `assert decoded_with != "hello"` instead of the length-based check.
